### PR TITLE
Update Zebra engine core from hoshir/zebra

### PR DESCRIPTION
**Pull request by oers**
## Summary

Syncs the vendored Zebra C engine core (`project/src/main/jni/zebra/`) from upstream [`hoshir/zebra`](https://github.com/hoshir) (reference `v1.2-hoshir`), following `docs/ZEBRA_ENGINE_UPDATE.md`, then validates the result on genuine arm64 hardware and fixes a long-standing native bug the validation work uncovered.

### Engine sync
- Refreshes the vendored Zebra sources from upstream, including the new parallel-search worker pool (`zebra/threads.c`/`threads.h`, registered in `Android.mk`). Kept single-threaded for now (`threads_init(1)` at JNI init) since the JNI bridge uses thread-bound `JNIEnv*` pointers and can't yet safely take callbacks from worker threads.
- Restores six local Android-fork defensive patches the raw upstream sync silently overwrote (verified none of them exist in the upstream tag either, so this isn't reverting anything upstream intentionally removed): a fuzzer-found `make_move()` occupied-square guard, a `disks_played` bounds guard in `unmake_move()`, the issue #82 crash fix in `inherit_move_lists()`, two array-bounds guards in `middle_game()`, a loaded-game validation check in `setup_game()`, and an `ILLEGAL`-move guard in `learn_game()` (plus a missing `#include` the last one needed to compile, and two more bounds-safety fixes in `setup_game()`'s token loop caught along the way).
- Re-applies the other Android-specific adaptations the update doc calls out (`android_files_dir`/`coeffs2.bin` handling, synchronous JNI eval/PV/status callbacks, etc.) on top of the refreshed sources.

### Validating the sync on real arm64 (not just x86 + ABI translation)
The existing CI only ever ran the engine on an x86 emulator with app-level ABI translation - it can't catch genuine ARM64-specific bugs, which is exactly what this sync's "ARM64-safe bitmasks/shifts" work is about. A real `arm64-v8a` emulator was tried first but never boots on GitHub-hosted runners (no `/dev/kvm`, confirmed against the long-open `android-emulator-runner#377`).

Landed instead on a host-native JNI approach: `project/hostjni/Makefile` compiles the same `droidzebra-jni.c` + zebra sources Android.mk does, but for the host toolchain instead of the NDK, so a plain JVM unit test (`HostJniSmokeTest`) can load the real engine and drive `ZebraEngine`'s actual public API directly - no Android runtime, no emulator, genuine native execution on whatever CPU the runner has. On `ubuntu-24.04-arm` that's real arm64, with zero virtualization needed. Getting the CI job itself green took several rounds (Android SDK bootstrapping, a Gradle lifecycle-task-vs-Test-task mismatch, and - the most interesting one - AGP's `aapt2` having no native arm64 build at all, resolved by pointing `android.aapt2FromMavenOverride` at a build matching this project's exact `buildToolsVersion`).

### The bug this surfaced: undo/redo across a forced pass
Bulk-replaying real WThor games under undo/redo reproduced a long-standing bug (previously tracked as `testRedoAcrossPass`, `@Suppress`'d in `WthorReplayTest#replayGamesWithUndoAndRedo` for years). Root-caused using the host-JNI harness as a fast (~seconds/iteration), no-emulator repro tool:

`_droidzebra_redo_turn` stopped exactly at its numeric `disks_played` target, even when that landed on a forced-pass position - unlike `_droidzebra_undo_turn`, which already symmetrically absorbs a pass into the same call. The trailing pass was instead left to the main game loop's own async auto-pass handling, which raced the *next* `redo()` call: if that arrived before the engine reached `ES_USER_INPUT_WAIT` again, it was silently dropped, and the real move right after the pass never got redone. Fixed by making `_droidzebra_redo_turn` absorb a trailing forced pass itself, symmetric with undo.

`WthorReplayTest#replayGamesWithUndoAndRedo`'s first-ever real CI run (after years of `@Suppress`) caught one more thing: its own `expectedColor` check assumed strict black/white ply alternation, which breaks for any game containing a pass - fixed to read the expected color off already-verified board data instead of re-deriving it from parity.

### Rebalancing test coverage: exhaustive corpus off the emulator, onto arm64
`WthorReplayTest`'s `MOVE_BY_MOVE`/`UNDO_AND_REDO` modes drove every one of ~200 WThor games through the real app on the emulator (550-700+s each) - but what they actually prove at that point is the Android integration layer (Activity/UI thread/GameStateBoardModel), not engine correctness, which doesn't care whether it's driven by a UI tap or a direct API call. That full-corpus emulator run was also the source of a CI flake hit while landing this PR (confirmed as a flake, not a real bug, by replaying all 200 games' full undo/redo cycles locally against the same engine with zero failures, and by the same commit going green on a subsequent CI run).

- `WthorReplayTest`: `MOVE_BY_MOVE`/`UNDO_AND_REDO` now run a small, curated set of 4 games (no pass, one forced pass - the exact bug-fix game, several passes, a short game ending early) instead of the full corpus; `replayGamesFast` keeps all ~200 (cheap, ~15s).
- `HostJniSmokeTest`: new `allWthorGamesUndoRedoRoundTripCleanly` takes over exhaustive-corpus coverage - full undo/redo round trip (whole-board diff at every ply) for up to 200 games, on real arm64 hardware, no emulator, ~250s total.

### Also covering the actual AI search, not just move application
Every `HostJniSmokeTest` scenario up to this point used `FUNCTION_HUMAN_VS_HUMAN`, which forces `skill=0` for both sides regardless of the depth passed to `EngineConfig` - so none of them ever called `compute_move()`: they exercised move application (`make_move`/`undo_turn`/`redo_turn`) but never the search itself (`middle_game`, `tree_search`/alpha-beta, bitboard move generation, pattern evaluation), which is exactly the code this sync's ARM64-safe bitmasks/shifts work touched. New `aiSearchPlaysLegalMovesAtVariousDepths` runs genuine AI-vs-AI games (`FUNCTION_ZEBRA_VS_ZEBRA`) at a few shallow depths (1, 3, 6) - different depths exercise different search code paths (iterative deepening steps, MPC re-searches, hash table usage) - and checks the game progresses with a self-consistent board. (Its first version raced the autoplaying background engine thread when reading the move count and the board a few lines apart; fixed by reading both back-to-back from a settled state instead, verified flake-free over 10 consecutive local runs.)

### Also fixed
- `full_pv`/`full_pv_depth` (search.c) made `_Thread_local`, matching `pv`/`pv_depth` and the rest of the search state the new worker-pool sync already thread-localizes. Not reachable today (single-threaded), but a landmine otherwise if that's ever turned on.

### Filed for later
[#100](https://github.com/oers/reversatile/issues/100) - the actual (separate, still open) "PV completion error" a fuzzer-found move-generation edge case already hinted at in the codebase (`moves.c`). Not caused by this sync; not fixed here; documented with the strongest lead found and a suggested fuzzing approach for whoever picks it up.

## Test plan
- [x] `externalNativeBuildDebug` succeeds for `arm64-v8a` (and all ABIs via `assembleDebug`) - CI `build`
- [x] `connectedDebugAndroidTest` (WThor regression suite, incl. `replayGamesWithUndoAndRedo`) - CI `test`
- [x] Host-native JNI test suite on genuine arm64 hardware, no emulator, including the full 200-game corpus and the new AI-search-depth coverage - CI `test-host-jni-arm64`
- [x] Manual local verification before every push: full 59-undo/59-redo round trip on the forced-pass game (5 repeated runs, no flakes); all 200 games' full undo/redo round trip (zero failures); AI-search-depth test (10 repeated runs, no flakes after the settle-based fix)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01V5G4cWSLBkiNNK4KC3zvXM

---
_Generated by [Claude Code](https://claude.ai/code/session_01V5G4cWSLBkiNNK4KC3zvXM)_

### CODE REVIEW

File: project/src/main/jni/zebra/moves.c [hunk 1]
```diff
@@ -220,16 +218,6 @@ game_in_progress( void ) {
 
 INLINE int
 make_move( int side_to_move, int move, int update_hash ) {
```
Comment (oers) [inline project/src/main/jni/zebra/moves.c:220]: **correctness (reachable):** `make_move()` dropped the defensive `board[move]==EMPTY` check that was added for a known fuzzer-found bug (the original comment cited a concrete repro: `-r 0 -l 9 6 3 5 19 0 ...`).

Without it, calling `make_move()` on an already-occupied square silently overwrites `board[move]` and updates `piece_count` as if a new disc was placed, desyncing board state from `disks_played` — previously this failed fast; now it can corrupt state and surface later as a confusing crash in PV completion instead.

Please restore the guard from before the upstream sync.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/moves.c [hunk 2]
```diff
@@ -330,9 +318,7 @@ make_move_no_hash( int side_to_move, int move ) {
 INLINE void
 unmake_move( int side_to_move, int move ) {
```
Comment (oers) [inline project/src/main/jni/zebra/moves.c:326]: **correctness (reachable):** `unmake_move()` dropped its `disks_played` bounds guard before decrementing. If it's ever called with `disks_played==0` (including as a downstream effect of the `make_move` guard being dropped above), `disks_played` goes to -1 and `hash_stored1[-1]`/`hash_stored2[-1]`/`flip_count[-1]` are read out of bounds.

Please restore the guard.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/search.c [hunk 3]
```diff
@@ -99,9 +101,7 @@ void
 inherit_move_lists( int stage ) {
```
Comment (oers) [inline project/src/main/jni/zebra/search.c:101]: **correctness (reachable) — reopens a previously-fixed crash (issue #82, commit 19091b2):** `inherit_move_lists()` lost the `stage >= 61 || stage < 0` bounds check before `list_inherited[stage] = TRUE`.

`midgame.c` calls `inherit_move_lists(disks_played + max_depth)` unconditionally, so whenever that sum reaches 61+ (a deep search late in the game), this writes past the end of the 61-element array and corrupts adjacent memory — this repo already hit and fixed exactly this crash once.

Please restore the bounds check.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/midgame.c [hunk 4]
```diff
@@ -1366,7 +1363,7 @@ middle_game( int side_to_move, int max_depth,
 
     /* Update the stored scores */
 
-    if ((base_stage + depth > 0 && base_stage + depth < 62) && (!stage_reached[base_stage + depth] || full_length_line) &&
+    if ( (!stage_reached[base_stage + depth] || full_length_line) &&
```
Comment (oers) [inline project/src/main/jni/zebra/midgame.c]: **correctness (reachable), same regression class as the `search.c` finding on this PR:** `middle_game()` dropped the `(base_stage + depth > 0 && base_stage + depth < 62)` bounds check before indexing `stage_reached[]`/`stage_score[]` (61-element arrays).

When `base_stage + depth` reaches 61+ or is ≤0, this writes/reads out of bounds of those thread-local arrays.

Please restore the check.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/osfbook.c [hunk 5]
```diff
@@ -4159,7 +4043,6 @@ fill_move_alternatives( int side_to_move,
 
     if ( child_feasible && (score == 0) &&
```
Comment (oers) [inline project/src/main/jni/zebra/osfbook.c:4044]: **correctness (reachable during normal book-enabled play):** `fill_move_alternatives()` dropped the `book_hash_table[slot] != EMPTY_HASH_SLOT` guard before using it to index `node[]`.

For a candidate move whose hash slot is empty/unavailable, this can now evaluate `node[book_hash_table[slot]].flags`, i.e. `node[-1].flags` — an out-of-bounds read.

Please restore the guard.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/game.c [hunk 6]
```diff
@@ -199,8 +201,7 @@ global_terminate( void ) {
 
 static void
 setup_game( const char *file_name, int *side_to_move ) {
```
Comment (oers) [inline project/src/main/jni/zebra/game.c:203]: **correctness (currently unreachable in the Android build, but worth restoring):** `setup_game()` dropped the `fatal_error` check rejecting a loaded game file whose center squares (d4/e4/d5/e5) aren't the standard Othello start — the removed comment says other code "is not ready for" that case.

Its only caller is in `zebra.c`, which isn't compiled into `Android.mk`, so this can't fire today. Flagging since the guard is gone from `game.c` itself and would bite the moment any caller passes a non-NULL `file_name`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/learn.c [hunk 7]
```diff
@@ -163,9 +156,6 @@ learn_game( int game_length, int private_game, int save_database ) {
       side_to_move = OPP( side_to_move );
       generate_all( side_to_move );
     }
-    if (game_move[i] == -1) {
-        fatal_error("Cannot learn game. Missing move no. %d", i);
-    }
     (void) make_move( side_to_move, game_move[i], TRUE );
```
Comment (oers) [inline project/src/main/jni/zebra/learn.c:163]: **correctness (currently unreachable in the Android build, but worth restoring):** `learn_game()` dropped the `fatal_error` check for an `ILLEGAL` (-1) stored move before calling `make_move()` with it, which would feed a negative index into board/hash arrays.

Its only caller is in `zebra.c`, not compiled into `Android.mk`, so unreachable today — but the safety net is gone from `learn.c` itself.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by oers — COMMENTED** [review]
Reviewed against `docs/ZEBRA_ENGINE_UPDATE.md`'s update process. The Android-specific restorations the doc's checklist calls out (`android_files_dir`/`coeffs2.bin` path handling, `clear_evaluated()`, `clear_endgame_performed()`, `get_stored_move()`, the synchronous JNI callbacks, `unpack_compressed_database_gz()`, ARM64-safe bitmasks, `threads_init(1)`, `threads.c` in `Android.mk`) all look correctly carried forward.

However, the raw upstream sync of `zebra/*.c` also silently dropped several local defensive patches that aren't on that checklist — two of them (`moves.c` make_move/unmake_move) are everyday-reachable, two more (`search.c`/`midgame.c` stage-array bounds) reopen a crash this repo already fixed once (issue #82), and one (`osfbook.c` book lookup) is reachable during normal book-enabled play. Two more are currently unreachable in the Android `.so` (their only callers live in `zebra.c`, which isn't in `Android.mk`) but should still be restored for safety. Left inline comments on each with specifics.

**Recommend not merging until these are restored** — several are real correctness/out-of-bounds regressions, not stylistic nits. Worth adding these patches to the update doc's checklist too, so the next sync doesn't drop them again.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/osfbook.c (hunk 5 above)
Comment (oers) [reply project/src/main/jni/zebra/osfbook.c:4044]: **Retracting this one — false positive.** I compared this against upstream `hoshir/zebra` at the `v1.2-hoshir` tag (the reference this PR's update doc names) and the `book_hash_table[slot] == EMPTY_HASH_SLOT` guard is present and correctly gates every subsequent `node[book_hash_table[slot]]` access here, identical to this PR's version (the `if (...) {...} else if ((node[book_hash_table[slot]].flags & flags) || !flags) {...}` chain). Line 4044 itself is actually `if (child_feasible && (score == 0) && ...)`, not an unguarded access. Sorry for the noise — resolving.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/moves.c (hunk 1 above)
Comment (oers) [reply project/src/main/jni/zebra/moves.c:220]: Fixed in 6586d89 — restored the guard verbatim from before the sync.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/moves.c (hunk 2 above)
Comment (oers) [reply project/src/main/jni/zebra/moves.c:326]: Fixed in 6586d89 — restored the guard verbatim from before the sync.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/search.c (hunk 3 above)
Comment (oers) [reply project/src/main/jni/zebra/search.c:101]: Fixed in 6586d89 — restored the bounds check verbatim from before the sync.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/midgame.c (hunk 4 above)
Comment (oers) [reply project/src/main/jni/zebra/midgame.c]: Fixed in 6586d89 — restored this guard, and also found+restored a second, related instance a few lines below (the read side of the same stage_reached[]/stage_score[] pair, in the oscillation-averaging block) that wasn't in my original finding.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/game.c (hunk 6 above)
Comment (oers) [reply project/src/main/jni/zebra/game.c:203]: Fixed in 6586d89 — restored the guard verbatim from before the sync.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: project/src/main/jni/zebra/learn.c (hunk 7 above)
Comment (oers) [reply project/src/main/jni/zebra/learn.c:163]: Fixed in 6586d89 — restored the guard verbatim from before the sync.

---
_Generated by [Claude Code](https://claude.ai/code)_
